File: tablet/Resources/task/Phq9.js
// Phq9.js

/*
    Copyright (C) 2012-2016 Rudolf Cardinal (rudolf@pobox.com).
    Department of Psychiatry, University of Cambridge.
    Funded by the Wellcome Trust.

    This file is part of CamCOPS.

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/

/*jslint node: true, newcap: true, nomen: true, plusplus: true */
"use strict";
/*global Titanium, L */

var DBCONSTANTS = require('common/DBCONSTANTS'),
    dbcommon = require('lib/dbcommon'),
    taskcommon = require('lib/taskcommon'),
    lang = require('lib/lang'),
    tablename = "phq9",
    fieldlist = dbcommon.standardTaskFields(),
    nquestions = 10,
    nscoredquestions = 9;

dbcommon.appendRepeatedFieldDef(fieldlist, "q", 1, nquestions,
                                DBCONSTANTS.TYPE_INTEGER);

// CREATE THE TABLE

dbcommon.createTable(tablename, fieldlist);

// TASK

function Phq9(patient_id) {
    taskcommon.BaseTask.call(this, patient_id); // call base constructor
}

lang.inheritPrototype(Phq9, taskcommon.BaseTask);
lang.extendPrototype(Phq9, {

    // KEY DATABASE FIELDS (USED BY DatabaseObject)

    _objecttype: Phq9,
    _tablename: tablename,
    _fieldlist: fieldlist,

    // TASK CLASS FIELD OVERRIDES (USED BY BaseTask)

    // OTHER

    // Scoring
    getTotalScore: function () {
        return taskcommon.totalScoreByPrefix(this, "q", 1, nscoredquestions);
    },

    // Standard task functions
    isComplete: function () {
        if (!taskcommon.isCompleteByPrefix(this, "q", 1, 9)) {
            return false;
        }
        if (this.q10 === null) {
            if (this.getTotalScore() === 0) {
                return true;
            }
            return false;
        }
        return true;
    },

    getSummary: function () {
        return (L('total_score') + " " + this.getTotalScore() + "/27" +
                this.isCompleteSuffix());
    },

    getDetail: function () {
        var uifunc = require('lib/uifunc'),
            totalscore = this.getTotalScore(),
            severity = (totalscore >= 20 ? L('severe')
                        : (totalscore >= 15 ? L('moderately_severe')
                           : (totalscore >= 10 ? L('moderate')
                              : (totalscore >= 5 ? L('mild')
                                 : L('none')
                                )
                             )
                          )
            ),
            ncore = (this.q1 >= 2 ? 1 : 0) +
                    (this.q2 >= 2 ? 1 : 0),
            nother = (this.q3 >= 2 ? 1 : 0) +
                     (this.q4 >= 2 ? 1 : 0) +
                     (this.q5 >= 2 ? 1 : 0) +
                     (this.q6 >= 2 ? 1 : 0) +
                     (this.q7 >= 2 ? 1 : 0) +
                     (this.q8 >= 2 ? 1 : 0) +
                     (this.q9 >= 1 ? 1 : 0), // Suicidality: counted WHENEVER present
            ntotal = ncore + nother,
            mds = (ncore >= 1) && (ntotal >= 5),
            ods = (ncore >= 1) && (ntotal >= 2) && (ntotal <= 4);
        // Scoring: ref PMID 10568646, http://www.phqscreeners.com/instructions/instructions.pdf
        return (
            taskcommon.valueDetail(this, "phq9_q", "_s", " ", "q", 1,
                                   nquestions) +
            "\n" +
            this.getSummary() + "\n" +
            "\n" +
            L('phq9_mds') + " " + uifunc.yesNo(mds) + "\n" +
            L('phq9_ods') + " " + uifunc.yesNo(ods) + "\n" +
            L('phq9_depression_severity') + " " + severity
        );
    },

    edit: function (readOnly) {
        var self = this,
            KeyValuePair = require('lib/KeyValuePair'),
            Questionnaire = require('questionnaire/Questionnaire'),
            pages,
            questionnaire;

        pages = [
            {
                title: L('phq9_title_main'),
                elements: [
                    { type: "QuestionText", text: L('phq9_stem'), bold: true },
                    {
                        type: "QuestionMCQGrid",
                        options: [
                            new KeyValuePair(L('phq9_a0'), 0),
                            new KeyValuePair(L('phq9_a1'), 1),
                            new KeyValuePair(L('phq9_a2'), 2),
                            new KeyValuePair(L('phq9_a3'), 3)
                        ],
                        questions: [
                            L('phq9_q1'),
                            L('phq9_q2'),
                            L('phq9_q3'),
                            L('phq9_q4'),
                            L('phq9_q5'),
                            L('phq9_q6'),
                            L('phq9_q7'),
                            L('phq9_q8'),
                            L('phq9_q9')
                        ],
                        fields: [
                            'q1', 'q2', 'q3', 'q4', 'q5',
                            'q6', 'q7', 'q8', 'q9'
                        ]
                    },
                    {
                        type: "QuestionText",
                        text: L('phq9_finalq'),
                        bold: true
                    },
                    {
                        type: "QuestionMCQ",
                        options: [
                            new KeyValuePair(L('phq9_fa0'), 0),
                            new KeyValuePair(L('phq9_fa1'), 1),
                            new KeyValuePair(L('phq9_fa2'), 2),
                            new KeyValuePair(L('phq9_fa3'), 3)
                        ],
                        field: 'q10'
                    }
                ]
            }
        ];

        questionnaire = new Questionnaire({
            readOnly: readOnly,
            pages: pages,
            callbackThis: self,
            fnGetFieldValue: self.defaultGetFieldValueFn,
            fnSetField: self.defaultSetFieldFn,
            fnFinished: self.defaultFinishedFn,
            fnShowNext: function () {
                return {
                    care: true,
                    showNext: self.isComplete()
                };
            }
        });
        questionnaire.open();
    }

});

module.exports = Phq9;

File: server/tasks/phq9.py
#!/usr/bin/env python3
# phq9.py

"""
    Copyright (C) 2012-2016 Rudolf Cardinal (rudolf@pobox.com).
    Department of Psychiatry, University of Cambridge.
    Funded by the Wellcome Trust.

    This file is part of CamCOPS.

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
"""

from cc_modules.cc_constants import (
    CTV_DICTLIST_INCOMPLETE,
)
from cc_modules.cc_db import repeat_fieldname, repeat_fieldspec
from cc_modules.cc_html import (
    answer,
    get_yes_no,
    tr,
    tr_qa,
)
from cc_modules.cc_string import WSTRING
from cc_modules.cc_task import get_from_dict, Task


# =============================================================================
# PHQ-9
# =============================================================================

class Phq9(Task):
    tablename = "phq9"
    shortname = "PHQ-9"
    longname = "Patient Health Questionnaire-9"
    fieldspecs = repeat_fieldspec(
        "q", 1, 9, min=0, max=3,
        comment_fmt="Q{n} ({s}) (0 not at all - 3 nearly every day)",
        comment_strings=[
            "anhedonia",
            "mood",
            "sleep",
            "energy",
            "appetite",
            "self-esteem/guilt",
            "concentration",
            "psychomotor",
            "death/self-harm",
        ]
    ) + [
        dict(name="q10", cctype="INT", min=0, max=3,
             comment="Q10 (difficulty in activities) (0 not difficult at "
                     "all - 3 extremely difficult)"),

    ]

    N_MAIN_QUESTIONS = 9

    def is_complete(self):
        if not self.field_contents_valid():
            return False
        if not self.are_all_fields_complete(repeat_fieldname("q", 1, 9)):
            return False
        if self.total_score() > 0 and self.q10 is None:
            return False
        return True

    def get_trackers(self):
        return [
            {
                "value": self.total_score(),
                "plot_label": "PHQ-9 total score (rating depressive symptoms)",
                "axis_label": "Score for Q1-9 (out of 27)",
                "axis_min": -0.5,
                "axis_max": 27.5,
                "axis_ticks": [
                    (27, "27"),
                    (25, "25"),
                    (20, "20"),
                    (15, "15"),
                    (10, "10"),
                    (5, "5"),
                    (0, "0"),
                ],
                "horizontal_lines": [
                    19.5,
                    14.5,
                    9.5,
                    4.5
                ],
                "horizontal_labels": [
                    (23, WSTRING("severe")),
                    (17, WSTRING("moderately_severe")),
                    (12, WSTRING("moderate")),
                    (7, WSTRING("mild")),
                    (2.25, WSTRING("none"))
                ]
            }
        ]

    def get_clinical_text(self):
        if not self.is_complete():
            return CTV_DICTLIST_INCOMPLETE
        return [{"content": "PHQ-9 total score {}/27 ({})".format(
            self.total_score(), self.severity())}]

    def get_summaries(self):
        return [
            self.is_complete_summary_field(),
            dict(name="total", cctype="INT", value=self.total_score(),
                 comment="Total score (/27)"),
            dict(name="n_core", cctype="INT", value=self.n_core(),
                 comment="Number of core symptoms"),
            dict(name="n_other", cctype="INT", value=self.n_other(),
                 comment="Number of other symptoms"),
            dict(name="n_total", cctype="INT", value=self.n_total(),
                 comment="Total number of symptoms"),
            dict(name="is_mds", cctype="BOOL", value=self.is_mds(),
                 comment="PHQ9 major depressive syndrome?"),
            dict(name="is_ods", cctype="BOOL", value=self.is_ods(),
                 comment="PHQ9 other depressive syndrome?"),
            dict(name="severity", cctype="TEXT", value=self.severity(),
                 comment="PHQ9 depression severity"),
        ]

    def total_score(self):
        return self.sum_fields(repeat_fieldname("q", 1, 9))

    def one_if_q_ge(self, qnum, threshold):
        value = getattr(self, "q" + str(qnum))
        return 1 if value is not None and value >= threshold else 0

    def n_core(self):
        return (self.one_if_q_ge(1, 2) +
                self.one_if_q_ge(2, 2))

    def n_other(self):
        return (self.one_if_q_ge(3, 2) +
                self.one_if_q_ge(4, 2) +
                self.one_if_q_ge(5, 2) +
                self.one_if_q_ge(6, 2) +
                self.one_if_q_ge(7, 2) +
                self.one_if_q_ge(8, 2) +
                self.one_if_q_ge(9, 1))  # suicidality
        # suicidality counted whenever present

    def n_total(self):
        return self.n_core() + self.n_other()

    def is_mds(self):
        return True if self.n_core() >= 1 and self.n_total() >= 5 else False

    def is_ods(self):
        return True if self.n_core() >= 1 and self.n_total() >= 2 and \
            self.n_total() <= 4 else False

    def severity(self):
        total = self.total_score()
        if total >= 20:
            return WSTRING("severe")
        elif total >= 15:
            return WSTRING("moderately_severe")
        elif total >= 10:
            return WSTRING("moderate")
        elif total >= 5:
            return WSTRING("mild")
        else:
            return WSTRING("none")

    def get_task_html(self):
        MAIN_DICT = {
            None: None,
            0: "0 — " + WSTRING("phq9_a0"),
            1: "1 — " + WSTRING("phq9_a1"),
            2: "2 — " + WSTRING("phq9_a2"),
            3: "3 — " + WSTRING("phq9_a3")
        }
        Q10_DICT = {
            None: None,
            0: "0 — " + WSTRING("phq9_fa0"),
            1: "1 — " + WSTRING("phq9_fa1"),
            2: "2 — " + WSTRING("phq9_fa2"),
            3: "3 — " + WSTRING("phq9_fa3")
        }
        h = """
            <div class="summary">
                <table class="summary">
        """ + self.get_is_complete_tr()
        h += tr(WSTRING("total_score") + " <sup>[1]</sup>",
                answer(self.total_score()) + " / 27")
        h += tr_qa(WSTRING("phq9_depression_severity") + " <sup>[2]</sup>",
                   self.severity())
        h += tr(
            "Number of symptoms: core <sup>[3]</sup>, other <sup>[4]</sup>, "
            "total",
            answer(self.n_core()) + "/2, "
            + answer(self.n_other()) + "/7, "
            + answer(self.n_total()) + "/9"
        )
        h += tr_qa(WSTRING("phq9_mds") + " <sup>[5]</sup>",
                   get_yes_no(self.is_mds()))
        h += tr_qa(WSTRING("phq9_ods") + " <sup>[6]</sup>",
                   get_yes_no(self.is_ods()))
        h += """
                </table>
            </div>
            <div class="explanation">
                Ratings are over the last 2 weeks.
            </div>
            <table class="taskdetail">
                <tr>
                    <th width="60%">Question</th>
                    <th width="40%">Answer</th>
                </tr>
        """
        for i in range(1, self.N_MAIN_QUESTIONS + 1):
            nstr = str(i)
            h += tr_qa(WSTRING("phq9_q" + nstr),
                       get_from_dict(MAIN_DICT, getattr(self, "q" + nstr)))
        h += tr_qa("10. " + WSTRING("phq9_finalq"),
                   get_from_dict(Q10_DICT, self.q10))
        h += """
            </table>
            <div class="footnotes">
                [1] Sum for questions 1–9.
                [2] Total score ≥20 severe, ≥15 moderately severe,
                    ≥10 moderate, ≥5 mild, &lt;5 none.
                [3] Number of questions 1–2 rated ≥2.
                [4] Number of questions 3–8 rated ≥2, or question 9
                    rated ≥1.
                [5] ≥1 core symptom and ≥5 total symptoms (as per
                    DSM-IV-TR page 356).
                [6] ≥1 core symptom and 2–4 total symptoms (as per
                    DSM-IV-TR page 775).
            </div>
        """
        return h
